2020/02/24/symfony-quickstart.html.md

```diff
@@ -0,0 +1,429 @@
+
+There is no ORM which solves everything, sometimes we want to write our own scripts. We can of course do it, but Doctrine offers [DQL](https://www.doctrine-project.org/projects/doctrine-orm/en/2.7/reference/dql-doctrine-query-language.html) (Doctrine Query Language) as a compromise.
+
+```php
+<?php
+$query = $em->createQuery('SELECT u FROM ForumUser u WHERE (u.username = :name OR u.username = :name2) AND u.id = :id');
+$query->setParameters(array(
+    'name' => 'Bob',
+    'name2' => 'Alice',
+    'id' => 321,
+));
+$users = $query->getResult(); // array of ForumUser objects
+```
+
+The example above is a parameterized query, where we search by username and id, without knowing what the actual filter values are at the time of development. These values are determined at runtime.
+
+One can apply entity changes upon the database via
+
+```
+php bin/console doctrine:migrations:migrate
+```
+
+Honestly I’m not very fond of changing the database schema based on entity classes, I consider this to be an anti-pattern, because we use a tool to generate a schema based on entities. Such a tool can have bugs, or, we might have errors in the entities. It is much better to do it the other way around, that is, make a proper database schema, write SQL commands to change the schema whenever we need it and generate entity classes that way. Luckily Doctrine offers that feature as well:
+
+```
+php bin/console doctrine:mapping:import "App\Entity" annotation --path=src/Entity
+```
+
+Unfortunately it’s a trend to write entity classes and generate schema changes based on those, but that way we add a layer of complexity around schema planning, which should be as simple as possible. If something goes south with these migrations for any reason, that could result in very serious problems, while if we are to write schema-changing scripts in SQL rather than generating them from entity classes that we write (or generate), then we at least have the means to acquire full understanding of what happens with the schema. Not everyone is a database specialist and most of the programmers are much more comfortable in writing or even generating entity classes, but even if we generate them, it’s not much simpler in most cases than writing a small script which creates a table or alters it or drops it, so the objective gain from generating schema-changing scripts from entity classes is negligible at best.
+
+On the other hand, the complex tool that we add is difficult to understand unless one delves into the actual code which generates the schema and the way the schema is generated is out of our hands, unless we are content with the current version of doctrine or accept the price of hacking into the code each time a new version is downloaded.
+
+In short, my advice is to try to avoid generating schema changes based on entity classes whenever possible. We can always find exceptional cases, but in general, SQL is clearer than a tool that generates SQL.
```

Comment:
@jonjensen I wanted to avoid it, so I have intended to emphasize the domain of the argument to make sure it will not be misunderstood. We of course need an ORM in many cases because that level of abstraction on data-level is saving a lot of time. However, in my opinion the line we should not cross is the schema scripting, because the problems we might face are much worse than the problems you have described earlier in the scenario when there is no rollback script in a case of urgency.

If, for example, for any reason the rollback script generated by Doctrine is not working, then we are not only in a situation of urgently resolve the issue with a live db problem, but we do not even know what we want to do.

So, I believe a nice compromise is to approach the issue in the following way:

- we keep schema scripting in our hands, maybe using some tools, but definitely in a manner that we fully understand what happens
- we include a step to generate rollback scripts at deploy time
- we use ORM for data management in most cases
- when the ORM is not smart-enough, we write some scripts on our own, but very carefully and keeping track where such custom scripts are in the source-code